src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -0,0 +1,353 @@
+#include <boost/algorithm/string/split.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/outcome/success_failure.hpp>
+
+#include <exception>
+
+namespace cloud_storage {
+
+static constexpr size_t max_cache_capacity_bytes = 1_GiB;
+
+materialized_manifest_cache::materialized_manifest_cache(
+  size_t capacity_bytes, retry_chain_logger& parent_logger)
+  : _capacity_bytes(capacity_bytes)
+  , _ctxlog(parent_logger)
+  , _sem(max_cache_capacity_bytes) {
+    vassert(
+      capacity_bytes > 0 && capacity_bytes < max_cache_capacity_bytes,
+      "Invalid cache capacity {}, should be non-zero and below 1GiB",
+      capacity_bytes);
+}
+
+ss::future<ss::semaphore_units<>> materialized_manifest_cache::prepare(
+  size_t size_bytes, std::optional<ss::lowres_clock::duration> timeout) {
+    ss::gate::holder h(_gate);
+    if (size_bytes > _capacity_bytes) {
+        vlog(
+          _ctxlog.trace,
+          "Oversized 'put' operation requested. Manifest size is {} bytes",
+          size_bytes);
+        // Oversized manifest handling. The manifest could be larger than
+        // capacity. If we will not allow this manifest to be added to the
+        // cache the subsystem will stall. The only possible solution is to
+        // let the manifest into the cache and allow it to evict everything
+        // else.
+        size_bytes = _capacity_bytes;
+    }
+    auto maybe_units = ss::try_get_units(_sem, size_bytes);
+    if (maybe_units.has_value()) {
+        vlog(
+          _ctxlog.trace,
+          "{} units acquired without waiting, {} available",
+          size_bytes,
+          _sem.available_units());
+        // The cache is not full and can grant some capacity without
+        // eviction
+        co_return std::move(maybe_units.value());
+    }
+    // The cache is full, try to free up some space. Free at least
+    // 'size_bytes' bytes.
+    size_t bytes_evicted = 0;
+    std::deque<model::offset> evicted;
+    while (bytes_evicted < size_bytes && !_cache.empty()) {
+        auto it = _access_order.begin();
+        auto so = it->manifest.get_start_offset();
+        vassert(
+          so.has_value(),
+          "Manifest can't be empty, ntp: {}",
+          it->manifest.get_ntp());
+        auto cit = _cache.find(so.value());
+        vassert(
+          cit != _cache.end(), "Manifest at {} already evicted", so.value());
+        evicted.push_back(so.value());
+        // Invariant: the materialized_manifest is always linked to either
+        // _access_order or _eviction_rollback list.
+        bytes_evicted += evict(cit, _eviction_rollback);
+    }
+    // Here the least recently used materialized manifests were evicted to
+    // free up 'size_bytes' bytes. But these manifests could still be used
+    // by some cursor. We need to wait for them to get released.
+    ss::semaphore_units<> u;
+    try {
+        if (timeout.has_value()) {
+            u = co_await ss::get_units(_sem, size_bytes, timeout.value());
```

Comment:
Using timeouts here is a bit risky: it means we won't strictly be respecting the order in which readers are trying to promote things into the cache, because under heavy load some of them will time out and clients will retry.

If the references to materialized manifests are going to be held by reader objects, maybe it makes sense to think about these things together: if we are waiting for something to release a manifest, then we are waiting for a reader to drop, so we have somehow ended up with more readers than we could handle manifests.  Maybe it makes sense to hook this together with `materialized_segments` as a central place for managing the concurrency, and we could proactively trim readers when we need to drop some manifests.

Reply:
In the next PR I have a `async_manifest_view_cursor` class which actually holds on to the shared pointer with the spillover manifest. This object has a built in timer and will release the pointer if it holds it for too long. So there will be a guarantee that the eviction will happen in time.